segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentStateStore.java

```diff
@@ -72,40 +89,136 @@
 
     @Override
     public CompletableFuture<SegmentState> get(String segmentName, Duration timeout) {
-        String stateSegment = StreamSegmentNameUtils.getStateSegmentName(segmentName);
+        return findLatestValidState(segmentName, timeout);
+    }
+
+    private CompletableFuture<SegmentState> findLatestValidState(String segmentName, Duration timeout) {
+        String stateSegment1 = StreamSegmentNameUtils.getFirstStateSegmentName(segmentName);
+        String stateSegment2 = StreamSegmentNameUtils.getSecondStateSegmentName(segmentName);
+
+        final CompletableFuture<SegmentState>[] states = new CompletableFuture[2];
+        final SegmentProperties[] props = new SegmentProperties[2];
+
         TimeoutTimer timer = new TimeoutTimer(timeout);
-        return this.storage
-                .getStreamSegmentInfo(stateSegment, timer.getRemaining())
-                .thenComposeAsync(sp -> readSegmentState(sp, timer.getRemaining()), this.executor)
+
+        CompletableFuture<SegmentState> retVal1 = this.storage
+                .getStreamSegmentInfo(stateSegment1, timer.getRemaining())
+                .thenComposeAsync(sp -> {
+                    props[0] = sp;
+                    states[0] = readSegmentState(sp, timer.getRemaining());
+                    return states[0];
+                }, this.executor)
+                .exceptionally(this::handleSegmentNotExistsException);
+
+        CompletableFuture<SegmentState> retVal2 = this.storage
+                .getStreamSegmentInfo(stateSegment2, timer.getRemaining())
+                .thenComposeAsync(sp -> {
+                    props[1] = sp;
+                    states[1] = readSegmentState(sp, timer.getRemaining());
+                    return states[1];
+                }, this.executor)
                 .exceptionally(this::handleSegmentNotExistsException);
+
+        return CompletableFuture.allOf(retVal1, retVal2).thenApplyAsync((v) -> {
```

Comment:
This is repeated in the method below. There should be a private method to return the newer one.

Reply:
There is a method to return the latest valid state and older/invalid statefilename. There is a lot of similar code but I am not able to refactor in such a way that it is reusable. Do you have any recommendations?

Reply:
While it's a little more work, I think the best thing to do would be to have SegmentState implement comparable with whatever sequence number etc we are going to use. Then you could add a function to MathHelpers in common for min and max that takes a comparable (Always choosing the non-null value)

A function like that would be generally useful. Looking at the code, I can spot at least three places that could be simplified if it existed: `LogMetadata:227`, `NewestReference:30`, and `StateSynchronizerImpl` on lines 149 and 186.